prelude.dx

```diff
@@ -256,16 +256,24 @@ def scan (init:a) (body:n->a->(a&b)) : (a & n=>b) =
     y
 
 def fold (init:a) (body:(n->a->a)) : a = fst $ scan init \i x. (body i x, ())
+def reduce (identity:a) (binop:(a->a->a)) (xs:n=>a) : a =
+  -- `binop` should be a commutative and associative, and form a
+  -- commutative monoid with `identity`
+  -- TODO: implement with a parallelizable monoid-parameterized writer
```

Comment:
I like the Monoid typeclass idea.  In the best case it could be neat and tidy when there's only one possible monoid, and the compiler would warn you whenever there's more than one possible monoid that could be used, is that correct?  Are there extremely general monoids that would make the implicit form useless because they would apply to everything?

There have been a few times so far when I've wished I could explicitly pass in an implicit argument (e.g. for ```randIdx``` and ```scan```.  I ended up making non-auto variants, which was unsatisfying and duplicated code.

Another use for being able to explicitly pass implicit arguments would be to allow overridable default arguments (like numerical tolerances).  I suppose one potential downside is that it might complicate things like currying.